### PR TITLE
[New] `jsx-no-target-blank`: add fixer

### PR DESCRIPTION
**Pull request by Nokel81**

### CODE REVIEW

File: lib/rules/jsx-no-target-blank.js [hunk 1]
```diff
@@ -103,7 +104,47 @@ module.exports = {
...
             message: 'Using target="_blank" without rel="noreferrer" '
-              + 'is a security risk: see https://html.spec.whatwg.org/multipage/links.html#link-type-noopener'
+              + 'is a security risk: see https://html.spec.whatwg.org/multipage/links.html#link-type-noopener',
+            fix(fixer) {
+              const relAttribute = node.parent.attributes.find((attr) => attr.name.name === 'rel');
+              if (!relAttribute) {
+                return fixer.insertTextAfter(node, ' rel="noreferrer"');
+              }
+
+              if (!relAttribute.value) {
```
Comment (ljharb) [inline lib/rules/jsx-no-target-blank.js:114]: can you add test cases for `rel=""`, `rel={false}`, `rel={0}`, etc?

File: lib/rules/jsx-no-target-blank.js [hunk 2]
```diff
@@ -103,7 +104,47 @@ module.exports = {
...
+                const parts = relAttribute.value.value
+                  .split('noreferrer')
+                  .filter((part) => part);
+                return fixer.replaceText(relAttribute.value, `"${parts.join(' ')} noreferrer"`);
+              }
+
+              if (relAttribute.value.expression.type === 'Literal') {
+                const valueType = typeof relAttribute.value.expression.value;
+                if (
+                  ['boolean', 'number', 'symbol', 'bigint', 'undefined'].includes(valueType)
```
Comment (ljharb) [inline lib/rules/jsx-no-target-blank.js:132]: what about `null`?

Comment (Nokel81) [reply lib/rules/jsx-no-target-blank.js:132]: That is caught by the `|| <expr>` on line 133 (right below this one). Remember, `typeof null` results in `"object"`.

File: lib/rules/jsx-no-target-blank.js (hunk 1 above)
Comment (Nokel81) [reply lib/rules/jsx-no-target-blank.js:114]: sure

Comment (Nokel81) [reply lib/rules/jsx-no-target-blank.js:114]: already existed for `""`. So I add for false, 0, and null

File: lib/rules/jsx-no-target-blank.js (hunk 2 above)
Comment (ljharb) [inline lib/rules/jsx-no-target-blank.js:132]: not a huge deal, but engines optimize the inline `typeof x === 'y'` pattern and do not optimize this pattern, so it'd likely be faster to do:
```suggestion
                if (
                  typeof relAttribute.value.expression.value === 'undefined'
                  || typeof relAttribute.value.expression.value === 'boolean'
                  || typeof relAttribute.value.expression.value === 'number'
                  || typeof relAttribute.value.expression.value === 'symbol'
                  || typeof relAttribute.value.expression.value === 'bigint'
```

File: lib/rules/jsx-no-target-blank.js [hunk 3]
```diff
@@ -103,7 +104,47 @@ module.exports = {
...
+                  || relAttribute.value.expression.value === null
+                ) {
+                  return fixer.replaceText(relAttribute.value.expression, '"noreferrer"');
+                }
+
+                if (valueType === 'string') {
+                  const parts = relAttribute.value.expression.value
+                    .split('noreferrer')
+                    .filter((part) => part);
```
Comment (ljharb) [inline lib/rules/jsx-no-target-blank.js:141]:
```suggestion
                    .filter(Boolean);
```

File: tests/lib/rules/jsx-no-target-blank.js [hunk 4]
```diff
@@ -96,60 +96,97 @@ ruleTester.run('jsx-no-target-blank', rule, {
...
     code: '<a target="_blank" rel="" href="http://example.com"></a>',
+    output: '<a target="_blank" rel="noreferrer" href="http://example.com"></a>',
+    errors: defaultErrors
+  }, {
+    code: '<a target="_blank" rel={0} href="http://example.com"></a>',
+    output: '<a target="_blank" rel={"noreferrer"} href="http://example.com"></a>',
+    errors: defaultErrors
+  }, {
+    code: '<a target="_blank" rel={false} href="http://example.com"></a>',
+    output: '<a target="_blank" rel={"noreferrer"} href="http://example.com"></a>',
```
Comment (ljharb) [inline tests/lib/rules/jsx-no-target-blank.js:111]:
```suggestion
    output: '<a target="_blank" rel="noreferrer" href="http://example.com"></a>',
```
we shouldn't include the curly braces for a string literal autofix.

File: tests/lib/rules/jsx-no-target-blank.js [hunk 5]
```diff
@@ -96,60 +96,97 @@ ruleTester.run('jsx-no-target-blank', rule, {
...
     code: '<a target="_blank" href="http://example.com"></a>',
+    output: '<a target="_blank" rel="noreferrer" href="http://example.com"></a>',
     errors: defaultErrors
   }, {
     code: '<a target="_blank" rel="" href="http://example.com"></a>',
+    output: '<a target="_blank" rel="noreferrer" href="http://example.com"></a>',
+    errors: defaultErrors
+  }, {
+    code: '<a target="_blank" rel={0} href="http://example.com"></a>',
+    output: '<a target="_blank" rel={"noreferrer"} href="http://example.com"></a>',
```
Comment (ljharb) [inline tests/lib/rules/jsx-no-target-blank.js:107]:
```suggestion
    output: '<a target="_blank" rel="noreferrer" href="http://example.com"></a>',
```

File: tests/lib/rules/jsx-no-target-blank.js [hunk 6]
```diff
@@ -96,60 +96,97 @@ ruleTester.run('jsx-no-target-blank', rule, {
...
+    errors: defaultErrors
+  }, {
+    code: '<a target="_blank" rel={false} href="http://example.com"></a>',
+    output: '<a target="_blank" rel={"noreferrer"} href="http://example.com"></a>',
+    errors: defaultErrors
+  }, {
+    code: '<a target="_blank" rel={null} href="http://example.com"></a>',
+    output: '<a target="_blank" rel={"noreferrer"} href="http://example.com"></a>',
```
Comment (ljharb) [inline tests/lib/rules/jsx-no-target-blank.js:115]:
```suggestion
    output: '<a target="_blank" rel="noreferrer" href="http://example.com"></a>',
```

File: tests/lib/rules/jsx-no-target-blank.js [hunk 7]
```diff
@@ -96,60 +96,97 @@ ruleTester.run('jsx-no-target-blank', rule, {
...
     code: '<a target="_blank" href="//example.com" rel={"noopenernoreferrer"}></a>',
+    output: '<a target="_blank" href="//example.com" rel={"noopener noreferrer"}></a>',
     errors: defaultErrors
   }, {
     code: '<a target={"_blank"} href={"//example.com"} rel={"noopenernoreferrer"}></a>',
+    output: '<a target={"_blank"} href={"//example.com"} rel={"noopener noreferrer"}></a>',
+    errors: defaultErrors
+  }, {
+    code: '<a target={"_blank"} href={"//example.com"} rel={"noopenernoreferrernoreferrernoreferrernoreferrernoreferrer"}></a>',
+    output: '<a target={"_blank"} href={"//example.com"} rel={"noopener noreferrer"}></a>',
```
Comment (ljharb) [inline tests/lib/rules/jsx-no-target-blank.js:158]: it's fine to keep it here since the original was both a string and had curly braces

**Review by ljharb — COMMENTED** [review]
Almost there :-)

File: lib/rules/jsx-no-target-blank.js (hunk 2 above)
Comment (Nokel81) [reply lib/rules/jsx-no-target-blank.js:132]: Ah thanks, didn't know they normally optimize that. Will do.

Comment (ljharb) [reply lib/rules/jsx-no-target-blank.js:132]: nobody does :-p it's super obscure knowledge

**Review by ljharb — APPROVED** [review]

**Comment by ljharb** [comment]
@Nokel81 ok! i think this time i've got it down to 4 failures, not just in older eslint versions: namely, that the fixer *is* fixing the code in the spread attribute case. If CI ends up showing those same 4 failures, then fixing that would be the last step to landing this.

**Comment by ljharb** [comment]
Please mark the PR as ready for review once either the tests are passing or you're stuck again (hopefully the former :-p )

**Comment by Nokel81** [comment]
@ljharb All tests are passing.

**Review by ljharb — APPROVED** [review]

**Pull request merged**